lib/net.c

```diff
@@ -1200,12 +1201,119 @@ start_iface(struct gatekeeper_if *iface, unsigned int num_attempts_link_get)
 	return ret;
 }
 
+static void
+calculate_net_config_para(struct net_config *net_conf)
+{
+	int i;
+
+	/* The total number of RX queues on any interface. */
+	uint16_t gatekeeper_num_rx_queues = net_conf->front.num_rx_queues +
+		(net_conf->back_iface_enabled ?
+		 net_conf->back.num_rx_queues : 0);
+
+	/* The total number of TX queues on any interface. */
+	uint16_t gatekeeper_num_tx_queues = net_conf->front.num_tx_queues +
+		(net_conf->back_iface_enabled ?
+		 net_conf->back.num_tx_queues : 0);
+
+	const char *filtered_blocks[] = { "dynamic_conf" };
+
+	/*
+	 * The total number of lcores used by any functional block
+	 * for either RX or TX except for Dynamic Configuration.
+	 */
+	int gatekeeper_num_lcores = launch_count_lcores(filtered_blocks,
+		RTE_DIM(filtered_blocks));
+
+	/*
+	 * The maximum number of receive descriptors to
+	 * allocate for the receive ring over all interfaces.
+	 */
+	uint16_t gatekeeper_num_rx_desc =
+		net_conf->front.gatekeeper_num_rx_desc;
+
+	/*
+	 * The maximum number of transmit descriptors to
+	 * allocate for the transmit ring over all interfaces.
+	 */
+	uint16_t gatekeeper_num_tx_desc =
+		net_conf->front.gatekeeper_num_tx_desc;
+
+	if (net_conf->back_iface_enabled) {
+		gatekeeper_num_rx_desc = RTE_MAX(gatekeeper_num_rx_desc,
+			net_conf->back.gatekeeper_num_rx_desc);
+
+		gatekeeper_num_tx_desc = RTE_MAX(gatekeeper_num_tx_desc,
+			net_conf->back.gatekeeper_num_tx_desc);
+	}
+
+	/*
+	 * The number of elements in the mbuf pool.
+	 *
+	 * The optimum size (in terms of memory usage) for a mempool is when
+	 * it is a power of two minus one.
+	 *
+	 * Need to provision enough memory for the worst case. According to
+	 * l3fwd sample application, it's essentially the number of RX
+	 * descriptors (across all queues for all ports), the number of TX
+	 * descriptors (across all queues for all ports), the number of packet
+	 * burst buffers (across all lcores for all bursts), the number of
+	 * slots in the packet buffer cache (across all lcores).
+	 */
+	net_conf->gatekeeper_num_mbuf = rte_align32pow2(
+		gatekeeper_num_rx_queues * gatekeeper_num_rx_desc +
+		net_conf->gatekeeper_num_ports *
+		net_conf->gatekeeper_total_pkt_burst +
```

Comment:
The l3fwd sample application doesn't have bonded interfaces, so the number of ports in that context is equivalent to the number of interfaces in Gatekeeper. Thus, we don't need the global parameter `net_conf->gatekeeper_num_ports` since it's only used here. We must replace `net_conf->gatekeeper_num_ports` here with a new local variable:
```
gatekeeper_num_if = net_conf->back_iface_enabled ? 2 : 1;
```

Reply:
In fact, if we replace `net_conf->gatekeeper_total_pkt_burst` with the following new parameters, we eliminate the last approximation in `net_conf->gatekeeper_num_mbuf`: `net_conf->gatekeeper_total_front_pkt_burst` and `net_conf->gatekeeper_total_back_pkt_burst`.

With these two new parameters, we can drop `gatekeeper_num_if` (mentioned in comment above) and replace the product `net_conf->gatekeeper_num_ports * net_conf->gatekeeper_total_pkt_burst` with `net_conf->gatekeeper_total_front_pkt_burst + net_conf->gatekeeper_total_back_pkt_burst`.

Reply:
Another insight. Instead of having `net_conf->gatekeeper_total_front_pkt_burst` and `net_conf->gatekeeper_total_back_pkt_burst`, add field `total_pkt_burst` to `struct gatekeeper_if`.